lib/puppet/parser/functions/upcase.rb

```diff
@@ -21,18 +21,18 @@ module Puppet::Parser::Functions
 
     value = arguments[0]
 
-    unless value.is_a?(Array) || value.is_a?(String) || value.is_a?(Hash)
+    unless value.is_a?(Array)  || value.is_a?(Hash) ||  value.respond_to?(:upcase)
       raise(Puppet::ParseError, 'upcase(): Requires an ' +
-                                  'array, string or hash to work with')
+                                  'array, hash or object that responds to upcase in order to work')
     end
 
     if value.is_a?(Array)
       # Numbers in Puppet are often string-encoded which is troublesome ...
-      result = value.collect { |i| i.is_a?(String) ? i.upcase : i }
+      result = value.collect { |i| function_upcase([i]) }
     elsif value.is_a?(Hash)
       result = {}
       value.each_pair do |k, v|
-        result.merge!({k.upcase => v.collect! { |p| p.upcase }})
+        result.merge!({function_upcase([k]) => function_upcase([v])})
```

Comment:
If you do mean `{'grant' => ['select'], 'Grant' => ['Update']}`, I have some bad news:
```
irb(main):009:0> {'grant'=>'select'}.merge({'grant'=> 'update'})
=> {"grant"=>"update"}
irb(main):010:0> {'grant'=>['select']}.merge({'grant'=> ['update']})
=> {"grant"=>["update"]}
irb(main):011:0> {'grant'=>{1=>'b'}}.merge({'grant'=> {2=>'c'}})
=> {"grant"=>{2=>"c"}}
```
It looks like `Hash#merge` overwrites old values with new ones too.